src/parse.cpp

```diff
@@ -818,11 +818,19 @@ expression* parser::parse_await_expression(token await_token, precedence prec) {
       });
     }
 
+    bool is_followed_by_async = this->peek().type == token_type::kw_async;
+
     expression* child = this->parse_expression(prec);
+
     if (child->kind() == expression_kind::_invalid) {
       this->error_reporter_->report(error_missing_operand_for_operator{
           .where = operator_span,
       });
+    } else if (this->in_async_function_ && this->is_arrow_kind(child) &&
+               !is_followed_by_async) {
```

Comment:
Nit: Instead of checking for the `async` token, you can ask if `child` is `async`:
```suggestion
               child->attributes() == function_attributes::async) {
```